### PR TITLE
[ML] Add log_time to AD data_counts and decide current based on it

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit is fixing a potential bug if we support anomaly detection
results index rollover in the future.

In particular, we determine the current `data_counts` by sorting on the
latest record time. However, this is not correct if the job reverts
to an older model snapshot. To fix this we add `log_time` to `data_counts`
(similarly to `model_size_stats`) and sort on `log_time` to figure
out the current counts for the job.

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/DataCounts.java [hunk 1]
```diff
@@ -138,31 +145,13 @@ public DataCounts(String jobId, long processedRecordCount, long processedFieldCo
         this.lastDataTimeStamp = lastDataTimeStamp;
         this.latestEmptyBucketTimeStamp = latestEmptyBucketTimeStamp;
         this.latestSparseBucketTimeStamp = latestSparseBucketTimeStamp;
+        this.logTime = logTime == null ? null : Instant.ofEpochMilli(logTime.toEpochMilli());
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/DataCounts.java:148]: unsure the rounding to epoch ms this is really needed. The only time that users will be using this object with actual data is when it is deserialized from the server.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java [hunk 2]
```diff
@@ -445,6 +445,10 @@ private SearchRequestBuilder createLatestDataCountsSearch(String indexName, Stri
                 .setIndicesOptions(IndicesOptions.lenientExpandOpen())
                 // look for both old and new formats
                 .setQuery(QueryBuilders.idsQuery().addIds(DataCounts.documentId(jobId), DataCounts.v54DocumentId(jobId)))
+                // We want to sort on log_time. However, this was added a long time later and before that we used to
+                // sort on latest_record_time. Thus we handle older data counts where no log_time exists and we fall back
+                // to the prior behaviour.
+                .addSort(SortBuilders.fieldSort(DataCounts.LOG_TIME.getPreferredName()).order(SortOrder.DESC).missing(0L))
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java:451]: Do we have to worry about `unmapped type` errors?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/DataCounts.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/process/DataCounts.java:148]: We need it for tests. We use `Instant.now()` in tests and that includes nanos. We do that everywhere we use `Instant`.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java:451]: Definitely. I tricked myself thinking `missing` was doing the job `unmapped` needs to do. Will fix that, thanks for spotting it!

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobResultsProvider.java:451]: Funny thing, I tested this. I created manually an index `.ml-anomalies-fake`. It all worked. But I didn't think that that index would pick our mappings due to the template being the new one.

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Pull request merged**